src/debugger/debug_workflow.ts

```diff
@@ -0,0 +1,156 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import { DBOSExecutor, DBOSNull, dbosNull } from "../dbos-executor";
+import { transaction_outputs } from "../../schemas/user_db_schema";
+import { IsolationLevel, Transaction, TransactionContext, TransactionContextImpl } from "../transaction";
+import { Communicator, CommunicatorContext, CommunicatorContextImpl } from "../communicator";
+import { DBOSDebuggerError, DBOSError, DBOSNotRegisteredError, DBOSWorkflowConflictUUIDError } from "../error";
+import { serializeError, deserializeError } from "serialize-error";
+import { SystemDatabase } from "../system_database";
+import { UserDatabaseClient } from "../user_database";
+import { SpanStatusCode } from "@opentelemetry/api";
+import { Span } from "@opentelemetry/sdk-trace-base";
+import { HTTPRequest, DBOSContext, DBOSContextImpl } from '../context';
+import { getRegisteredOperations } from "../decorators";
+import { WFInvokeFuncs, Workflow, WorkflowConfig, WorkflowContext, WorkflowHandle } from "../workflow";
+
+interface RecordedResult<R> {
+  output: R;
+  txn_snapshot: string;
+  txn_id: string;
+}
+
+/**
+ * Context used for debugging a workflow
+ */
+export class WorkflowContextDebug extends DBOSContextImpl implements WorkflowContext {
+  functionID: number = 0;
+  readonly #wfe;
+  readonly isTempWorkflow: boolean;
+
+  constructor(wfe: DBOSExecutor, parentCtx: DBOSContextImpl | undefined, workflowUUID: string, readonly workflowConfig: WorkflowConfig, workflowName: string) {
+    const span = wfe.tracer.startSpan(workflowName, { workflowUUID: workflowUUID }, parentCtx?.span);
+    super(workflowName, span, wfe.logger, parentCtx);
+    this.workflowUUID = workflowUUID;
+    this.#wfe = wfe;
+    this.isTempWorkflow = wfe.tempWorkflowName === workflowName;
+    if (wfe.config.application) {
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+      this.applicationConfig = wfe.config.application;
+    }
+  }
+
+  functionIDGetIncrement(): number {
+    return this.functionID++;
+  }
+
+  invoke<T extends object>(object: T): WFInvokeFuncs<T> {
+    const ops = getRegisteredOperations(object);
+
+    const proxy: any = {};
+    for (const op of ops) {
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+      proxy[op.name] = op.txnConfig
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+        ? (...args: any[]) => this.transaction(op.registeredFunction as Transaction<any[], any>, ...args)
+        : op.commConfig
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+        ? (...args: any[]) => this.external(op.registeredFunction as Communicator<any[], any>, ...args)
+        : undefined;
+    }
```

Comment:
@devhawk: would it be possible for the proxy to be built only once ? Rather than doing it at every `invoke`?

Reply:
We talked about using the same tech from openapi work to also do stuff like pre-generate invokers. But I would like to see some perf number suggesting this is an issue first